File: gs_operateur_mobile/lib/models/recu_perception.dart
class RecuPerception {
  final int id;
  final String numero;
  final int taxeId;
  final int? personneId;
  final bool valide;
  final String typePerception;
  final String dateEmission;
  final String? heureEmission;
  final String? categorieVehicule;
  final String? plaqueImmatriculation;
  final double montant;
  final String? trajet;
  final String? chauffeurNom;
  final String? conducteurNom;
  final String? numeroPiece;
  final String? designation;
  final double? poids;
  final String? observations;
  final int percepteurId;
  final String? createdAt;
  final TaxeRef? taxe;
  final PersonneSimple? personne;
  final PersonneSimple? percepteur;

  static const Map<String, String> typeLabels = {
    'peage_urbain': 'Péage urbain',
    'pont_bascule': 'Pont bascule',
    'etalage': 'Étalage',
    'chargement': 'Chargement',
    'dechargement': 'Déchargement',
    'autre': 'Autre',
  };

  static const Map<String, String> typePrefixes = {
    'peage_urbain': 'PEA',
    'pont_bascule': 'PON',
    'etalage': 'ETA',
    'chargement': 'CHA',
    'dechargement': 'DEC',
    'autre': 'AUT',
  };

  RecuPerception({
    required this.id,
    required this.numero,
    required this.taxeId,
    this.personneId,
    this.valide = false,
    required this.typePerception,
    required this.dateEmission,
    this.heureEmission,
    this.categorieVehicule,
    this.plaqueImmatriculation,
    required this.montant,
    this.trajet,
    this.chauffeurNom,
    this.conducteurNom,
    this.numeroPiece,
    this.designation,
    this.poids,
    this.observations,
    required this.percepteurId,
    this.createdAt,
    this.taxe,
    this.personne,
    this.percepteur,
  });

  factory RecuPerception.fromJson(Map<String, dynamic> json) {
    return RecuPerception(
      id: json['id'] ?? 0,
      numero: json['numero'] ?? '',
      taxeId: json['taxe_id'] ?? 0,
      personneId: json['personne_id'],
      valide: json['valide'] ?? false,
      typePerception: json['type_perception'] ?? '',
      dateEmission: json['date_emission'] ?? '',
      heureEmission: json['heure_emission'],
      categorieVehicule: json['categorie_vehicule'],
      plaqueImmatriculation: json['plaque_immatriculation'],
      montant: _toDouble(json['montant']),
      trajet: json['trajet'],
      chauffeurNom: json['chauffeur_nom'],
      conducteurNom: json['conducteur_nom'],
      numeroPiece: json['Numero_Piece'],
      designation: json['designation'],
      poids: _toDoubleNullable(json['poids']),
      observations: json['observations'],
      percepteurId: json['percepteur_id'] ?? 0,
      createdAt: json['created_at'],
      taxe: json['taxe'] != null ? TaxeRef.fromJson(json['taxe']) : null,
      personne: json['personne'] != null ? PersonneSimple.fromJson(json['personne']) : null,
      percepteur: json['percepteur'] != null ? PersonneSimple.fromJson(json['percepteur']) : null,
    );
  }

  static double _toDouble(dynamic v) {
    if (v == null) return 0;
    if (v is num) return v.toDouble();
    return double.tryParse(v.toString()) ?? 0;
  }

  static double? _toDoubleNullable(dynamic v) {
    if (v == null) return null;
    if (v is num) return v.toDouble();
    return double.tryParse(v.toString());
  }

  String get typeLabel => typeLabels[typePerception] ?? typePerception;
  String getNumero() => numero;
  bool get isVehicle => typePerception == 'peage_urbain' || typePerception == 'pont_bascule';
  bool get isMarchandise => typePerception == 'chargement' || typePerception == 'dechargement';
}

class TaxeRef {
  final int id;
  final String nom;
  final double? montant;

  TaxeRef({required this.id, required this.nom, this.montant});

  factory TaxeRef.fromJson(Map<String, dynamic> json) {
    return TaxeRef(
      id: json['id'] ?? 0,
      nom: json['nom'] ?? '',
      montant: _parseDouble(json['montant'] ?? json['taux']),
    );
  }

  static double? _parseDouble(dynamic v) {
    if (v == null) return null;
    if (v is num) return v.toDouble();
    return double.tryParse(v.toString());
  }
}

class PersonneSimple {
  final int id;
  final String nom;
  final String? prenom;
  final String? postnom;

  PersonneSimple({required this.id, required this.nom, this.prenom, this.postnom});

  factory PersonneSimple.fromJson(Map<String, dynamic> json) {
    return PersonneSimple(
      id: json['id'] ?? 0,
      nom: json['nom'] ?? json['nom_utilisateur'] ?? '',
      prenom: json['prenom'],
      postnom: json['postnom'],
    );
  }

  String get displayName {
    final parts = [nom, postnom, prenom].where((s) => s != null && s.isNotEmpty);
    return parts.join(' ');
  }
}

File: id_operateur/app/Models/RecuPerception.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class RecuPerception extends Model
{
    use HasFactory;

    protected $table = 'recus_perception';

    protected $fillable = [
        'taxe_id',
        'personne_id',
        'percepteur_id',
        'valide',
        'type_perception',
        'numero',
        'date_emission',
        'heure_emission',
        'categorie_vehicule',
        'plaque_immatriculation',
        'montant',
        'trajet',
        'chauffeur_nom',
        'conducteur_nom',
        'Numero_Piece',
        'designation',
        'poids',
        'observations',
    ];

    protected $casts = [
        'date_emission' => 'date',
        'montant' => 'decimal:2',
        'poids' => 'decimal:2',
        'valide' => 'boolean',
    ];

    // Relations
    public function taxe() { return $this->belongsTo(Taxe::class); }
    public function personne() { return $this->belongsTo(Personne::class); }
    public function percepteur() { return $this->belongsTo(Utilisateur::class, 'percepteur_id'); }

    // Types de perception
    public const TYPES = [
        'peage_urbain' => 'Péage urbain',
        'pont_bascule' => 'Pont bascule',
        'etalage' => 'Étalage',
        'chargement' => 'Chargement',
        'dechargement' => 'Déchargement',
        'autre' => 'Autre',
    ];

    // Générer le prochain numéro de série
    public static function prochainNumero(string $type = 'PEA'): string
    {
        $prefixe = strtoupper($type);
        $dernier = self::where('numero', 'like', "{$prefixe}/%")
            ->orderBy('id', 'desc')
            ->first();

        $numero = $dernier ? intval(substr($dernier->numero, strlen($prefixe) + 1)) + 1 : 1;
        return $prefixe . '/' . str_pad($numero, 3, '0', STR_PAD_LEFT);
    }

    public function getTrajetLabelAttribute(): string
    {
        $labels = [
            'aller' => 'Aller',
            'retour' => 'Retour',
            'aller_retour' => 'Aller/Retour',
        ];
        return $labels[$this->trajet] ?? $this->trajet;
    }

    public function getTypePerceptionLabelAttribute(): string
    {
        return self::TYPES[$this->type_perception] ?? $this->type_perception;
    }
}
